tools/testbench/common_test_ipc4.c

```diff
@@ -0,0 +1,593 @@
+
+#if defined __XCC__
+#include <xtensa/tie/xt_timer.h>
+#endif
+
+SOF_DEFINE_REG_UUID(testbench);
+DECLARE_TR_CTX(testbench_tr, SOF_UUID(testbench_uuid), LOG_LEVEL_INFO);
+LOG_MODULE_REGISTER(testbench, CONFIG_SOF_LOG_LEVEL);
+
+/* testbench helper functions for pipeline setup and trigger */
+
+int tb_setup(struct sof *sof, struct testbench_prm *tp)
+{
+	struct ll_schedule_domain domain = {0};
+	int bits;
+	int krate;
+
+	domain.next_tick = tp->tick_period_us;
+
+	/* init components */
+	sys_comp_init(sof);
+
+	/* Module adapter components */
+	sys_comp_module_crossover_interface_init();
+	sys_comp_module_dcblock_interface_init();
+	sys_comp_module_demux_interface_init();
+	sys_comp_module_drc_interface_init();
+	sys_comp_module_eq_fir_interface_init();
+	sys_comp_module_eq_iir_interface_init();
+	sys_comp_module_file_interface_init();
+	sys_comp_module_gain_interface_init();
+	sys_comp_module_google_rtc_audio_processing_interface_init();
+	sys_comp_module_igo_nr_interface_init();
+	sys_comp_module_mfcc_interface_init();
+	sys_comp_module_multiband_drc_interface_init();
+	sys_comp_module_mux_interface_init();
+	sys_comp_module_rtnr_interface_init();
+	sys_comp_module_selector_interface_init();
+	sys_comp_module_src_interface_init();
+	sys_comp_module_asrc_interface_init();
+	sys_comp_module_tdfb_interface_init();
+	sys_comp_module_volume_interface_init();
+
+	/* other necessary initializations, todo: follow better SOF init */
+	pipeline_posn_init(sof);
+	init_system_notify(sof);
+
+	/* init IPC */
+	if (ipc_init(sof) < 0) {
+		fprintf(stderr, "error: IPC init\n");
+		return -EINVAL;
+	}
+
+	/* Trace */
+	ipc_tr.level = LOG_LEVEL_INFO;
+	ipc_tr.uuid_p = SOF_UUID(testbench_uuid);
+
+	/* init LL scheduler */
+	if (scheduler_init_ll(&domain) < 0) {
+		fprintf(stderr, "error: edf scheduler init\n");
+		return -EINVAL;
+	}
+
+	/* init EDF scheduler */
+	if (scheduler_init_edf() < 0) {
+		fprintf(stderr, "error: edf scheduler init\n");
+		return -EINVAL;
+	}
+
+	tb_debug_print("ipc and scheduler initialized\n");
+
+	/* setup IPC4 audio format */
+	tp->num_configs = 1;
+	krate = tp->fs_in / 1000;
+	switch (tp->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		bits = 16;
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+		bits = 24;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		bits = 32;
+		break;
+	default:
+		fprintf(stderr, "error: unsupported frame format %d\n", tp->frame_fmt);
+		return -EINVAL;
+	}
+
+	/* TODO 44.1 kHz like rates */
+	snprintf(tp->config[0].name, TB_MAX_CONFIG_NAME_SIZE, "%dk%dc%db",
+		 krate, tp->channels_in, bits);
+	tp->num_configs = 1;
+	tp->config[0].buffer_frames = 2 * krate;
+	tp->config[0].buffer_time = 0;
+	tp->config[0].period_frames = krate;
+	tp->config[0].period_time = 0;
+	tp->config[0].rate = tp->fs_in;
+	tp->config[0].channels = tp->channels_in;
+	tp->config[0].format = tp->frame_fmt;
+	tp->period_size = 2 * krate;
+
+	/* TODO used for what? determine from topology and enabled pipelines if need */
```

Comment:
@lgirdwood do you know what this is for?

Reply:
It's a feature from SOF plugin, I'm not sure I need this. Maybe need if I with testbench -pA,B,C,.. command line switch select pipelines those are not connected to 1st PCM. I need to test with other topologies, e.g. sof-hda-generic that has more pipelines.

Reply:
@ranj063 do we still need this as an open or can we remove the comment/fix ?

Reply:
@singalsu with the plugin, the pipelines are determined based on just the PCM ID for the device that we want to start playback/capture on. I suppose this is not needed for the testbench because we only have 1 pipeline in your test topology and you want to set up all the widgets anyway.

Reply:
Thanks for explanation! I'll keep it for now for possible more complex scenarios. But I will need to update the PCM ID based on the command line specified pipelines.